### ISSUE TITLE
Add Categorical.from_codes

### ISSUE THREAD

**Issue by TomAugspurger** [dask/dask#2829]
I had a need for this today. I think something like

```python
In [74]: categories = ['a', 'b', 'c']

In [75]: codes = da.random.randint(0, 4, size=100, chunks=50)

In [76]: (dd.from_dask_array(codes)
    ...:    .astype('category')
    ...:    .cat.set_categories(np.arange(len(categories)))
    ...:    .cat.rename_categories(categories))
    ...:
    ...:
Out[76]:
Dask Series Structure:
npartitions=2
0     category[known]
50                ...
99                ...
dtype: category
Dask Name: cat, 10 tasks
```

will work (may not be optimal though). The more awkward part is finding where to put that, since dask doesn't have a `Categorical`, so maybe just a top-level function?

**Comment by ryan-williams** [comment]
The example code here really helped me, thanks. Some search-engine breadcrumbs:

This is a way to convert an existing Dask Series to a categorical dtype, letting the user provide the categories explicitly, and avoiding having to do a full pass over the data / `.compute()` (as is required by e.g. `dd.categorize()`).

**Comment by arnold-smit** [comment]
Thanks for helping me out with the code above.

I suspect this situation occurs quite often when importing data from relational databases with models using foreign keys into lookup / dimension tables. 

IMHO an ideal place to construct the Categorical would be in the dask.dataframe.read_<...> family of functions. 

Would love to have syntax like:
```python
dask.dataframe.read_csv(..., dtypes={'var': {'category', dict_from_codes_to_categories}, ...}) 
```
The validity of the dict could be checked immediately, and an exception could be raised when some illegal code is read on the fly. Have no idea on the complexity of this, but thought I'd share my thoughts. If someone points me in the right direction, I could have an attempt at contributing.

**Comment by TomAugspurger** [comment]
I don't think we would want that kind of codes -> categories behavior without adding it in pandas first. And my *guess* is that it wouldn't be accepted there.

With https://github.com/dask/dask/issues/5001, we have another motivation for a simple extension array collection. I'm guessing that will be implemented soon rather than later, and so we'll have a `dask.dataframe.Categorical.from_codes`.

**Comment by mrocklin** [comment]
> I don't think we would want that kind of codes -> categories behavior without adding it in pandas first. And my guess is that it wouldn't be accepted there.

I'm confused.  Isn't this the behavior of the Pandas `from_codes` function?

**Comment by TomAugspurger** [comment]
Should have been clearer. I meant the codes -> categories *in read_csv* that was proposed in https://github.com/dask/dask/issues/2829#issuecomment-591361411 is likely not going to happen.
